plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -212,10 +194,123 @@ public static Collection<QueryGroup> getFromClusterStateMetadata(String name, Cl
+            }
+        });
+    }
+
+    /**
+     * Modify cluster state to update the existing QueryGroup
+     * @param updateQueryGroupRequest {@link QueryGroup} - the QueryGroup that we want to update
+     * @param currentState - current cluster state
+     */
+    ClusterState updateQueryGroupInClusterState(UpdateQueryGroupRequest updateQueryGroupRequest, ClusterState currentState) {
+        final Metadata metadata = currentState.metadata();
+        final Map<String, QueryGroup> existingGroups = currentState.metadata().queryGroups();
+        String name = updateQueryGroupRequest.getName();
+
+        final QueryGroup existingGroup = existingGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(name))
+            .findFirst()
+            .orElseThrow(() -> new ResourceNotFoundException("No QueryGroup exists with the provided name: " + name));
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        if (updateQueryGroupRequest.getResourceLimits() != null && !updateQueryGroupRequest.getResourceLimits().isEmpty()) {
+            validateTotalUsage(existingGroups, name, updateQueryGroupRequest.getResourceLimits());
+        }
+
+        // build the QueryGroup with updated fields
+        final Map<ResourceType, Double> updatedResourceLimits = new HashMap<>(existingGroup.getResourceLimits());
+        if (updateQueryGroupRequest.getResourceLimits() != null) {
+            updatedResourceLimits.putAll(updateQueryGroupRequest.getResourceLimits());
+        }
+        final ResiliencyMode mode = Optional.ofNullable(updateQueryGroupRequest.getResiliencyMode())
+            .orElse(existingGroup.getResiliencyMode());
+
+        final QueryGroup updatedGroup = new QueryGroup(
+            name,
+            existingGroup.get_id(),
+            mode,
+            updatedResourceLimits,
+            updateQueryGroupRequest.getUpdatedAtInMillis()
+        );
+        return ClusterState.builder(currentState)
+            .metadata(Metadata.builder(metadata).remove(existingGroup).put(updatedGroup).build())
+            .build();
+    }
+
+    /**
+     * This method checks if there's any resource allocation that exceed limit of 1.0
+     * @param existingQueryGroups - existing QueryGroups in the system
+     * @param resourceLimits - the QueryGroup we're creating or updating
+     */
+    private void validateTotalUsage(Map<String, QueryGroup> existingQueryGroups, String name, Map<ResourceType, Double> resourceLimits) {
+        final Map<ResourceType, Double> totalUsage = new EnumMap<>(ResourceType.class);
+        totalUsage.putAll(resourceLimits);
+        for (QueryGroup currGroup : existingQueryGroups.values()) {
+            if (!currGroup.getName().equals(name)) {
+                for (ResourceType resourceType : resourceLimits.keySet()) {
+                    totalUsage.compute(resourceType, (k, v) -> v + currGroup.getResourceLimits().get(resourceType));
```

Comment:
What if the `v` in the compute lambda is `null` ? It could be the case when one of the existing query group does not have the resource type which `resourceLimits` contains.

Reply:
good catch